RecoLocalCalo/Castor/src/RecHitCorrector.cc

```diff
@@ -114,72 +112,49 @@ RecHitCorrector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
    CastorChannelQuality* myqual = new CastorChannelQuality(*p.product());
    
-   if (!rechits.isValid()) std::cout << "No valid CastorRecHitCollection found, please check the InputLabel..." << std::endl;
+   if (!rechits.isValid()) edm::LogWarning("CastorRecHitCorrector") << "No valid CastorRecHitCollection found, please check the InputLabel...";
    
    CastorCalibrations calibrations;
    
-   std::auto_ptr<CastorRecHitCollection> rec(new CastorRecHitCollection);
+   auto rec = std::make_unique<CastorRecHitCollection>();
    
    for (unsigned int i=0;i<rechits->size();i++) {
    	CastorRecHit rechit = (*rechits)[i];
-	//std::cout << "rechit energy = " << rechit.energy() << std::endl;
-	double fC = factor_*rechit.energy();
 	double time = rechit.time();
-	//std::cout << "rechit energy(fC) = " << fC << " time = " << time << std::endl;
-	
-	// do proper gain calibration reading the latest entries in the condDB
-	const CastorCalibrations& calibrations=conditions->getCastorCalibrations(rechit.id());
-	int capid = 0; // take some capid, gains are the same for all capid's
+	double correctedenergy = factor_*rechit.energy();
 	
-	double correctedenergy = 0;
 	if (doInterCalib_) {
-		if (rechit.id().module() <= 2) {
-			correctedenergy = 0.5*fC*calibrations.gain(capid);
-			//std::cout << " correctedenergy = " << correctedenergy << " gain = " << calibrations.gain(capid) << std::endl;
-		} else {
-			correctedenergy = fC*calibrations.gain(capid);
-		}
-	} else {
-		if (rechit.id().module() <= 2) {
-			correctedenergy = 0.5*fC;
-		} else {
-			correctedenergy = fC;
-		}
+		// do proper gain calibration reading the latest entries in the condDB
+		const CastorCalibrations& calibrations=conditions->getCastorCalibrations(rechit.id());
+		int capid = 0; // take some capid, gains are the same for all capid's
+		correctedenergy *= calibrations.gain(capid);
 	}
 	
 	// now check the channelquality of this rechit
 	bool ok = true;
 	DetId detcell=(DetId)rechit.id();
 	std::vector<DetId> channels = myqual->getAllChannels();
-	//std::cout << "number of specified quality flags = " << channels.size() << std::endl;
-	for (std::vector<DetId>::iterator channel = channels.begin();channel !=  channels.end();channel++) {	
-		if (channel->rawId() == detcell.rawId()) {
-			const CastorChannelStatus* mydigistatus=myqual->getValues(*channel);
-			//std::cout << "CastorChannelStatus = " << mydigistatus->getValue() << std::endl;
-			if (mydigistatus->getValue() == 2989) ok = false; // 2989 = BAD
+	for (auto channel : channels) {	
+		if (channel.rawId() == detcell.rawId()) {
+			const CastorChannelStatus* mydigistatus=myqual->getValues(channel);
+			if (mydigistatus->getValue() == 2989) {
+				ok = false; // 2989 = BAD
+				break;
+			}
 		}
 	}
 	
 	if (ok) {
-	    CastorRecHit *correctedhit = new CastorRecHit(rechit.id(),correctedenergy,time);
-	    rec->push_back(*correctedhit);
+	    rec->emplace_back(rechit.id(),correctedenergy,time);
```

Comment:
@hvanhaev , is it correct? I would think about "rec->emplace_back(new CastorRecHit(....))"

Reply:
No, using `new` here is totally unnecessary and in fact it's a memory leak. `rec` is a `CastorRecHitCollection` and it can construct the `CastorRecHit` objects from arguments using `emplace_back()`.